src/rpc/core_rpc_server.cpp

```diff
@@ -486,7 +486,16 @@ namespace cryptonote
           {
             pool_tx_hashes.insert(tx_hash);
             missed_txs.erase(mi);
-            txs.push_back(*i);
+            const std::vector<crypto::hash>::iterator iplace = std::find(vh.begin(), vh.end(), tx_hash);
+            if (iplace == vh.end())
+            {
+              res.status = "Internal error";
+              return true;
+            }
+            size_t idx = std::distance(vh.begin(), iplace);
+            std::list<transaction>::iterator j = txs.begin();
+            std::advance(j, idx);
+            txs.insert(j, *i);
```

Comment:
This seems problematic, as the `idx` may go above the range of `txs`; e.g. consider a case where the `req.txs_hashes` looks like this: `[<nonexistent_txhash_0>, <nonexistent_txhash_1>, <pool_txhash>, <mined_txhash>]`. When this line is executed, `idx` would be 2 but the size of `txs` would be 1, so `std::advance(j, idx)` would make the iterator `j` invalid, leading to an undefined behavior (even though it doesn't crash for most platforms since the `std::list` iterator is most often implemented in a cyclic manner, see [here](https://stackoverflow.com/questions/12992007/is-it-safe-to-use-stdlist-as-a-circular-list); it does crash if `txs` were declared as `std::vector`).

I'd implement it in the following way (sorry for a long paste):

```c++
    std::list<crypto::hash> missed_txs;
    std::list<transaction> mined_txs;
    bool r = m_core.get_transactions(vh, mined_txs, missed_txs);
    if(!r)
    {
      res.status = "Failed";
      return true;
    }
    std::unordered_map<crypto::hash, transaction> mined_txs_map;
    for (const auto& tx : mined_txs)
      mined_txs_map.insert(std::make_pair(get_transaction_hash(tx), tx));
    LOG_PRINT_L2("Found " << mined_txs.size() << "/" << vh.size() << " transactions on the blockchain");

    // try the pool for any missing txes
    // size_t found_in_pool = 0;
    std::unordered_map<crypto::hash, transaction> pool_txs_map;
    if (!missed_txs.empty())
    {
      std::list<transaction> pool_txs;
      bool r = m_core.get_pool_transactions(pool_txs);
      if(r)
      {
        for (std::list<transaction>::const_iterator i = pool_txs.begin(); i != pool_txs.end(); ++i)
        {
          crypto::hash tx_hash = get_transaction_hash(*i);
          std::list<crypto::hash>::iterator mi = std::find(missed_txs.begin(), missed_txs.end(), tx_hash);
          if (mi != missed_txs.end())
          {
            pool_txs_map.insert(std::make_pair(tx_hash, *i));
            missed_txs.erase(mi);
            const std::vector<crypto::hash>::iterator iplace = std::find(vh.begin(), vh.end(), tx_hash);
            if (iplace == vh.end())
            {
              res.status = "Internal error";
              return true;
            }
            // size_t idx = std::distance(vh.begin(), iplace);
            // std::list<transaction>::iterator j = txs.begin();
            // std::advance(j, idx);
            // txs.insert(j, *i);
            // ++found_in_pool;
          }
        }
      }
      LOG_PRINT_L2("Found " << pool_txs_map.size() << "/" << vh.size() << " transactions in the pool");
    }

    // merge txs from blockchain & pool while preserving the ordering
    std::vector<transaction> txs;
    for (const auto& tx_hash : vh)
    {
      const auto found_mined = mined_txs_map.find(tx_hash);
      if (found_mined != mined_txs_map.end())
      {
        txs.push_back(found_mined->second);
        continue;
      }
      const auto found_pool = pool_txs_map.find(tx_hash);
      if (found_pool != pool_txs_map.end())
        txs.push_back(found_pool->second);
    }

    // std::list<std::string>::const_iterator txhi = req.txs_hashes.begin();
    // std::vector<crypto::hash>::const_iterator vhi = vh.begin();
    for(auto& tx: txs)
    {
      res.txs.push_back(COMMAND_RPC_GET_TRANSACTIONS::entry());
      COMMAND_RPC_GET_TRANSACTIONS::entry &e = res.txs.back();

      crypto::hash tx_hash = get_transaction_hash(tx);
      e.tx_hash = string_tools::pod_to_hex(tx_hash);
      blobdata blob = t_serializable_object_to_blob(tx);
      e.as_hex = string_tools::buff_to_hex_nodelimer(blob);
      if (req.decode_as_json)
        e.as_json = obj_to_json_str(tx);
      e.in_pool = pool_txs_map.find(tx_hash) != pool_txs_map.end();
      if (e.in_pool)
      {
        e.block_height = std::numeric_limits<uint64_t>::max();
      }
      else
      {
        e.block_height = m_core.get_blockchain_storage().get_db().get_tx_block_height(tx_hash);
      }

      // fill up old style responses too, in case an old wallet asks
      res.txs_as_hex.push_back(e.as_hex);
      if (req.decode_as_json)
        res.txs_as_json.push_back(e.as_json);

      // output indices too if not in pool
      if (pool_txs_map.find(tx_hash) == pool_txs_map.end())
      {
        bool r = m_core.get_tx_outputs_gindexs(tx_hash, e.output_indices);
        if (!r)
        {
          res.status = "Failed";
          return false;
        }
      }
    }
```